### DISCUSSION TITLE
Help with decyphering RDKit Fingerprint

### DISCUSSION THREAD

**Discussion by Supervitux** [rdkit/rdkit#6889] [Q&A] [answered]
Dear community,

I am trying to create a graphic to explain the generation of the RDKit fingerprint. For that I tried to understand the hashing and the generation of random numbers after. I also wanted to come up with a realistic example, but I was not able to find any source for what a "hash" or a random number seeded by it looks like. Can anyone give me examples for those? 
Also can I find anywhere some details how the random numbers determine which bit is set to 1? I tried to figure it out from the source code (./Code/GraphMol/Fingerprints/RDKitFPGenerator.cpp ?) but I do not really know C++.
 Any help is appreciated! Here is the graphic, so far I just put some random numbers for the hash and the random numbers:
 ![TopFPExample](https://github.com/rdkit/rdkit/assets/46672499/d1a82d0d-1249-46e1-ae57-d713a35ba7d4)

**Comment by dehaenw** [comment, marked as answer]
it does not matter exactly which hashing function is used, what matters is that it has correct properties: it will convert some input (like a path in this case) to a hash, which in this case is 32-bit integer and in such a way that even slightly different inputs will give a totally different hash (to minimize the chance of collisions: the situation of 2 different inputs having the same hash). Folding these numbers into a fingerprint is usually just done by a modulo operation, with the modulus set to the length of the fingerprint. The resulting number is then used as the index of the bit you set to one.

In case you want to play around with what a hash looks like, the easiest way to do this is just using the hash() function inbuilt in python on the paths represented as strings. (hash() produces a 64 bit signed integer but the same principle applies).
```python
paths=["OC","CC","C=O","OCC","CC=O","OCC=O"]
hashes=[hash(path) for path in paths]
hashes
```
this gives 
```
[-132746954262126347, 1186591322618830029, 7297015890975028230, 7553121090342264500, -3940444275122232980,4647500731298589481]
```
(you may have different values, the output depends on the PYTHONHASHSEED env variable)

then, folding these values into a bit vector of length 16 would work as follows:
```python
fp=[0]*16
for h in hashes:
    fp[h%len(fp)]=1
print("".join([str(b) for b in fp]))
```
this results in the following fingerprint: 0000111001001100

**Reply by Supervitux** [reply]
Thank you very much for the detailed explanation. It is exactly what I needed.
